core/sail/inferencer/src/main/java/org/eclipse/rdf4j/sail/inferencer/fc/ForwardChainingSchemaCachingRDFSInferencer.java

```diff
@@ -0,0 +1,318 @@
+
+        try {
+           while(true) {
+                long l = readWriteLock.tryConvertToWriteLock(connection.lockStamp);
+
+                if (l != 0) {
+//                    long temp = connection.lockStamp;
+                    connection.lockStamp = l;
+//                    if (temp != l) {
+//                        System.err.println("readLock: " + temp + " writeLock: " + connection.lockStamp);
+//                    }
+
+
+                    return;
+                }
+
+                try {
+                    Thread.sleep(random.nextInt(2));
+                } catch (InterruptedException e) {
+                    // ignore interrupted exception
+                }
+
+                // detect potential deadlock scenario
+                if(numberOfThreadsWaitingForWriteLock.get() > 1  //More than 1 connection waiting for a write lock
+                    && numberOfThreadsWaitingForWriteLock.get() <= readWriteLock.getReadLockCount() // no connection only using a read lock
+                    && random.nextBoolean()){ // randomly kick this connection out
+                    break;
+                }
+
+            }
+
+//            System.err.println("isReadLocked: " + readWriteLock.isReadLocked());
+//            System.err.println("isWriteLocked(): " + readWriteLock.isWriteLocked());
+//            System.err.println("read lock count: " + readWriteLock.getReadLockCount());
+
+            releaseLock(connection);
+            throw new SailConflictException("Concurrent modification of schema, could not acquire the lock.");
+        } finally {
+
+            numberOfThreadsWaitingForWriteLock.decrementAndGet();
+        }
+
+    }
+
+
+    public ForwardChainingSchemaCachingRDFSInferencer(NotifyingSail data) {
+        super(data);
+        schema = null;
+
+    }
+
+
+    public ForwardChainingSchemaCachingRDFSInferencer(NotifyingSail data, Repository schema) {
+        super(data);
+
+        this.schema = schema;
+
+    }
+
+    public ForwardChainingSchemaCachingRDFSInferencer(NotifyingSail data, boolean useAllRdfsRules) {
+        super(data);
+        schema = null;
+
+        this.useAllRdfsRules = useAllRdfsRules;
+
+    }
+
+    public ForwardChainingSchemaCachingRDFSInferencer(NotifyingSail data, Repository schema, boolean useAllRdfsRules) {
+        super(data);
+
+        this.schema = schema;
+        this.useAllRdfsRules = useAllRdfsRules;
+
+    }
+
+
+    public void initialize() throws SailException {
+        super.initialize();
+
+        if (sharedSchema) {
+            return;
+        }
+
+
+        ForwardChainingSchemaCachingRDFSInferencerConnection connection = null;
+
+        try {
+            connection = getConnection();
+            final ForwardChainingSchemaCachingRDFSInferencerConnection finalConnection = connection;
+            finalConnection.begin();
+
+            finalConnection.addAxiomStatements();
+
+
+            List<Statement> tboxStatments = new ArrayList<>();
```

Comment:
If this is always going to be empty in the schema == null cases using Collections.emptyList() would be more appropriate for maintenance in the future to signal its intentional emptiness in the default case.